src/mito2/src/cache/index.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use api::v1::index::InvertedIndexMetas;
+use async_trait::async_trait;
+use common_base::BitVec;
+use index::inverted_index::error::DecodeFstSnafu;
+use index::inverted_index::format::reader::InvertedIndexReader;
+use index::inverted_index::FstMap;
+use prost::Message;
+use snafu::ResultExt;
+use uuid::Uuid;
+
+use crate::metrics::{CACHE_BYTES, CACHE_HIT, CACHE_MISS};
+
+/// Metrics for index metadata.
+const INDEX_METADATA_TYPE: &str = "index_metadata";
+/// Metrics for index content.
+const INDEX_CONTENT_TYPE: &str = "index_content";
+
+/// Inverted index blob reader with cache.
+pub struct CachedInvertedIndexBlobReader<R> {
+    file_id: Uuid,
+    inner: R,
+    cache: InvertedIndexCacheRef,
+}
+
+impl<R> CachedInvertedIndexBlobReader<R> {
+    pub fn new(file_id: Uuid, inner: R, cache: InvertedIndexCacheRef) -> Self {
+        Self {
+            file_id,
+            inner,
+            cache,
+        }
+    }
+}
+
+impl<R> CachedInvertedIndexBlobReader<R>
+where
+    R: InvertedIndexReader,
+{
+    /// Gets given range of index data from cache, and loads from source if the file
+    /// is not already cached.
+    async fn get_or_load(
+        &mut self,
+        offset: u64,
+        size: u32,
+    ) -> index::inverted_index::error::Result<Vec<u8>> {
+        let range = offset as usize..(offset + size as u64) as usize;
+        if let Some(cached) = self.cache.get_index(IndexKey {
+            file_id: self.file_id,
+        }) {
+            CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+            Ok(cached[range].to_vec())
+        } else {
+            let mut all_data = Vec::with_capacity(1024 * 1024);
+            self.inner.read_all(&mut all_data).await?;
+            let result = all_data[range].to_vec();
+            let key = IndexKey {
+                file_id: self.file_id,
+            };
+            let value = Arc::new(all_data);
+            let size = index_content_weight(&key, &value);
+            self.cache.put_index(key, value);
+            CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+            CACHE_BYTES
+                .with_label_values(&[INDEX_CONTENT_TYPE])
+                .add(size.into());
```

Comment:
Looks like `put_index()` already added `CACHE_BYTES`.

Reply:
> Looks like `put_index()` already added `CACHE_BYTES`.

fixed in f1e3558bd17e67e7a3ffe3df1f8cbf5ca9e44ccd